src/forder.c

```diff
@@ -1420,3 +1465,249 @@ SEXP binary(SEXP x)
   return ans;
 }
 
+// all(x==1L)
+static bool all1(SEXP x) {
+  if (!isInteger(x))
+    error("internal error: all1 got non-integer"); // # nocov
+  int *xp = INTEGER(x);
+  for (int i=0; i<LENGTH(x); ++i) if (xp[i] != 1) return false;
+  return true;
+}
+
+// identical(cols, head(chmatch(key(x), names(x)), length(cols)))
+bool colsKeyHead(SEXP x, SEXP cols) {
+  if (!isNewList(x))
+    error("internal error: 'x' must be a list"); // # nocov
+  if (!isInteger(cols))
+    error("internal error: 'cols' must be an integer"); // # nocov
+  SEXP key = getAttrib(x, sym_sorted);
+  if (isNull(key) || (length(key) < length(cols)))
+    return false;
+  SEXP names =  getAttrib(x, R_NamesSymbol);
+  SEXP keynames = PROTECT(chmatch(key, names, 0));
+  int *keynamesp = INTEGER(keynames), *colsp = INTEGER(cols);
```

Comment:
WDYT?

```suggestion
  SEXP key = PROTECT(getAttrib(x, sym_sorted));
  if (isNull(key) || (length(key) < length(cols))) {
    UNPROTECT(1); // key
    return false;
  }
  SEXP keynames = PROTECT(chmatch(key, getAttrib(x, R_NamesSymbol), 0));
  UNPROTECT(1); // key
  int *keynamesp = INTEGER(keynames), *colsp = INTEGER(cols);
```

Reply:
should be ok, but assuming `x` won't be gc'ed then we don't have anything to worry about, no?

Reply:
It basically makes sense, but I'm given pause by R itself doing `PROTECT(getAttrib(.))`:

https://github.com/search?q=repo%3Ar-devel%2Fr-svn%20%2FPROTECT%5B(%5DgetAttrib%2F&type=code

We also do it ourselves elsewhere, as do other CRAN packages:

https://github.com/search?q=org%3Acran+%2FPROTECT%5B%28%5DgetAttrib%2F&type=code

I think it's OK to err on the conservative side.